sycl/include/CL/sycl/types.hpp

```diff
@@ -2279,6 +2279,99 @@ __SYCL_DECLARE_FLOAT_VECTOR_CONVERTERS(double)
 #undef __SYCL_DECLARE_SCALAR_BOOL_CONVERTER
 #undef __SYCL_USE_EXT_VECTOR_TYPE__
 
+/// This macro must be defined to 1 when SYCL implementation allows user
+/// applications to explicitly declare certain class types as device copyable
+/// by adding specializations of is_device_copyable type trait class.
+#define SYCL_DEVICE_COPYABLE 1
+
+/// is_device_copyable is a user specializable class template to indicate
+/// that a type T is device copyable, which means that SYCL implementation
+/// may copy objects of the type T between host and device or between two
+/// devices.
+/// Specializing is_device_copyable such a way that
+/// is_device_copyable_v<T> == true on a T that does not satisfy all
+/// the requirements of a device copyable type is undefined behavior.
+template <typename T, typename = void>
+struct is_device_copyable : std::false_type {};
+
+template <typename T>
+struct is_device_copyable<
+    T, std::enable_if_t<std::is_trivially_copyable<T>::value>>
+    : std::true_type {};
+
+#if __cplusplus >= 201703L
+template <typename T>
+inline constexpr bool is_device_copyable_v = is_device_copyable<T>::value;
+#endif // __cplusplus >= 201703L
+
+// std::tuple<> is implicitly device copyable type.
+template <> struct is_device_copyable<std::tuple<>> : std::true_type {};
+
+// std::tuple<Ts...> is implicitly device copyable type if each type T of Ts...
+// is device copyable.
+template <typename T, typename... Ts>
+struct is_device_copyable<std::tuple<T, Ts...>>
+    : detail::bool_constant<is_device_copyable<T>::value &&
+                            is_device_copyable<std::tuple<Ts...>>::value> {};
+
+namespace detail {
+template <typename T, typename = void>
+struct IsDeprecatedDeviceCopyable : std::false_type {};
+
+// TODO: using C++ attribute [[deprecated]] or the macro __SYCL2020_DEPRECATED
+// does not produce expected warning message for the type 'T'.
+template <typename T>
+struct __SYCL2020_DEPRECATED("This type isn't device copyable in SYCL 2020")
+    IsDeprecatedDeviceCopyable<
+        T, std::enable_if_t<std::is_trivially_copy_constructible<T>::value &&
+                            std::is_trivially_destructible<T>::value &&
+                            !is_device_copyable<T>::value>> : std::true_type {};
```

Comment:
Am I understanding this correctly? We are saying an object `IsDeprecatedDeviceCopyable` if it is `trivially_copy_constructible` and `trivially_destructible` but NOT `trivially_copyable`  ?  If that's what we want to say, maybe we should use `!std::is_trivially_copyable`  so it's clearer.   
But if you wanted this to be "in the case where we are not device_copyable",  then the old way would suffice. But in that case, the OR checks will capture that.  If that is the desire, then maybe leave this third clause out of the conjunction.  
Please correct me if I misunderstood.

Reply:
> We are saying an object IsDeprecatedDeviceCopyable if it is trivially_copy_constructible and trivially_destructible but NOT trivially_copyable

That is exactly what is needed and meant by this code. If type is NOT trivially copyable, but is constructible/destructible, then it is allowed to be passed from HOST to DEVICE, but as deprecated behavior (to stay compatible with older SYCL std).

Reply:
Then I'd recommend using `!std::is_trivially_copyable` so it's easier to understand, and not tied to `is_device_copyable` in case that were to change in the future.

Reply:
is_device_copyable includes 2 situations: a) is_trivially_device_copyable, b) situation when user declared some type as device_copyable. Both of (a) and (b) are not considered deprecated pass for copying data from host to device. (a) and (b) are good ways (acc to SYCL 2020) to allow type copying.

Reply:
But in this case, wouldn't the  `is_device_copyable<T> || isDeprecatedDeviceCopyable<T>` capture that? It works even if `isDeprecatedDeviceCopyable` does not contain `!is_device_copyable`.    And that is the only use for `isDeprecatedDeviceCopyable`, to appear in those OR clauses. 
 Anyway, doing all these template substitutions and boolean logic in my head hurts. If only I had a machine that could do it for me.... ;-)